File: backend/app/adapters/llm/ollama.py
import httpx

from app.config import settings


async def is_ollama_available() -> bool:
    base = (settings.ollama_base_url or "").rstrip("/")
    if not base:
        return False
    try:
        async with httpx.AsyncClient(timeout=2.0) as client:
            res = await client.get(f"{base}/api/tags")
            return res.status_code == 200
    except httpx.HTTPError:
        return False


async def ollama_generate_json(prompt: str) -> dict:
    base = settings.ollama_base_url.rstrip("/")
    model = settings.ollama_model
    async with httpx.AsyncClient(timeout=120.0) as client:
        res = await client.post(
            f"{base}/api/generate",
            json={
                "model": model,
                "prompt": prompt,
                "stream": False,
                "format": "json",
            },
        )
        res.raise_for_status()
        body = res.json()
        text = (body.get("response") or "").strip()
        if not text:
            raise ValueError("empty ollama response")
        import json

        return json.loads(text)

File: conversate/web/src/lib/llm/ollama.ts
const DEFAULT_BASE = "http://127.0.0.1:11434";
const DEFAULT_MODEL = "gemma2:9b";

export function ollamaConfig(): { baseUrl: string; model: string } {
  return {
    baseUrl: (process.env.OLLAMA_BASE_URL ?? DEFAULT_BASE).replace(/\/$/, ""),
    model: process.env.OLLAMA_MODEL ?? DEFAULT_MODEL,
  };
}

export async function isOllamaAvailable(): Promise<boolean> {
  const { baseUrl } = ollamaConfig();
  try {
    const res = await fetch(`${baseUrl}/api/tags`, {
      signal: AbortSignal.timeout(2500),
      cache: "no-store",
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function ollamaGenerateJson<T>(opts: {
  prompt: string;
  parse: (raw: unknown) => T;
}): Promise<{ data: T; model: string; latencyMs: number }> {
  const { baseUrl, model } = ollamaConfig();
  const start = Date.now();
  const res = await fetch(`${baseUrl}/api/generate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model,
      prompt: opts.prompt,
      stream: false,
      format: "json",
    }),
    signal: AbortSignal.timeout(120_000),
  });
  if (!res.ok) {
    throw new Error(`ollama ${res.status}`);
  }
  const body = (await res.json()) as { response?: string };
  const text = body.response?.trim();
  if (!text) throw new Error("ollama empty response");
  const parsed = JSON.parse(text) as unknown;
  return {
    data: opts.parse(parsed),
    model,
    latencyMs: Date.now() - start,
  };
}
